vllm/model_executor/models/aquila.py

```diff
@@ -296,11 +297,18 @@ def forward(
         kv_caches: List[KVCache],
         input_metadata: InputMetadata,
         cache_events: Optional[List[torch.cuda.Event]],
-    ) -> SamplerOutput:
+    ) -> torch.Tensor:
         hidden_states = self.model(input_ids, positions, kv_caches,
                                    input_metadata, cache_events)
+        return hidden_states
+
+    def sample(
+        self,
+        hidden_states: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+    ) -> SamplerOutput:
         next_tokens = self.sampler(self.lm_head.weight, hidden_states,
-                                   input_metadata)
+                                   sampling_metadata)
         return next_tokens
```

Comment:
It seems like this part should be the same for all models. Can we move this part out of the model? This can be helpful if we want to implement different kinds of sampler in the future, for example speculative decoding sampler.

Reply:
That's a good question. There are actually subtle differences:

1. Some models tie the last head with the embedding layer, while others don't.
2. GPT-J and Phi models have embedding bias, while others don't.

Given these differences, I think it's a bit better to define the `sample` method for each model.

Reply:
One possible approach could be to implement an `BaseModel` class that implements the most common sampling approach, and have all the models inherit from it and override the method if necessary. On the other hand, explicitly defining `sample` in every model has its advantages too, as it allows the developer to immediately see the logic used for the particular model

Reply:
@zhuohan123 I believe the current design is conceptually simpler. Also, we will capture the forward method into CUDA graph (or torch.compile in the future). While your proposal also works with CUDA graph, having the `forward` method produce a single output tensor will make things simpler.

Reply:
@Yard1 Makes sense. I don't have strong preference between the current design and your proposal.

Reply:
I wouldn't block on this - we can do it in a follow up if needed.

Reply:
Sounds good!